django/contrib/admin/views/autocomplete.py

```diff
@@ -0,0 +1,92 @@
+from django.http import Http404, JsonResponse
+from django.utils.encoding import force_text
+from django.views.generic.list import BaseListView
+
+
+class AutocompleteJsonView(BaseListView):
+    """
+    View that handles requests from `.AutocompleteWidget`.
+
+    The view only supports HTTP's GET method.
+    """
+    paginate_by = 20
+    admin_site = None
+    source_model = None
+    source_model_admin = None
+
+    @property
+    def target_model(self):
+        related_field = self.source_model._meta.get_field(self.field_name)
+        return related_field.remote_field.model
+
+    @property
+    def target_model_admin(self):
+        try:
+            return self.admin_site._registry[self.target_model]
+        except KeyError:
+            raise Http404('%s is not registered in the admin.' % self.target_model)
+
+    def has_perm(self, request):
+        """Check if user has permission to access the related model."""
+        return self.source_model_admin.has_change_permission(request)
+
+    def get_paginator(self, queryset, per_page, orphans=0,
+                      allow_empty_first_page=True, **kwargs):
+        return self.target_model_admin.get_paginator(self.request, queryset, per_page, orphans, allow_empty_first_page)
+
+    def get(self, request, *args, **kwargs):
```

Comment:
Do we really want this to be GET requests or are POST with CSRF better suited. Thinking about DoS here :-/